### DISCUSSION TITLE
Fast table driven crc in forth

### DISCUSSION THREAD

**Discussion by Whammo** [jkotlinski/durexforth#430] [Ideas]
```forth
\ fast table driven crc in forth

: ftab 8 0 do dup 1 and if
1 rshift $8408 xor \ CRC-16-CCITT Polynomial
else 1 rshift then loop ;

: tomem $ff 0 do
i ftab $c000 \ address of table
i 2* + ! loop ;

tomem

: crctab 2* $c000 + @ ; \ return table entry

: crc_each ( crc c -- newcrc  )     \ init crc to $ffff
over xor $ff and crctab             \ calculate on each char of string 
swap 8 rshift xor ;

: output_crc ( crc -- msb lsb )     \ this returns crc  
dup invert 8 rshift
swap invert $ff and ;

: calc_crc ( addr length -- crc )
over c@ $ffff  \ addr length char $ffff \ init
swap crc_each  \ addr length crc
swap           \ addr crc length
1 do over      \ addr crc addr
i + c@         \ addr crc char  
crc_each loop  \ addr crc
nip output_crc ;
```

**Comment by burnsauce** [comment]
Cool!  I have a framework I've been using to check for collisions, I'll see how this fares.

**Reply by Whammo** [reply]
It's missing something..

 Nah- it's okay tomem one word

**Reply by burnsauce** [reply]
Okay, let me know when it's ready.  Alternately I can share my collision check code.

**Comment by burnsauce** [comment]
I might be missing something here, is this a 32-bit crc?  I'm getting a double as the output of `calc_crc`. Or is a 16-bit crc 32-bits long?

**Comment by Whammo** [comment]
Treat them as bytes. None should be over FF

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Reply by burnsauce** [reply]
Can I just combine them together into a 16-bit int instead?

**Comment by Whammo** [comment]
You may  but this is a special checksum.
Run the string, and save the checksum
 Then run the string again followed by the individual two bytes of the saved checksum
 That checksum will be the same as any unchanged strings checksum. $f0b8

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by Whammo** [comment]
Just change the two addresses as you please.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by burnsauce** [comment]
Unfortunately this will not work as a hash.  `lda,` and `included` produce the same crc.

Good code for error detection though!

**Reply by Whammo** [reply]
It's the PPP CRC Rfc 1661

**Comment by Whammo** [comment]
You might have better results if you pad the strings to a fixed length with a repeatable sequence.

**Reply by burnsauce** [reply]
Is there a magic string that makes for good padding?  Whatever I try isn't helping.

**Reply by burnsauce** [reply]
Wait, no I had a bug.

**Comment by burnsauce** [comment]
There's going to have to be *some* magic in the padded data.

Unsurprisingly, the crc of `lit` and `litc` are the same when the pad string is `AbBcCdD...`

Unprintable characters maybe?

**Reply by burnsauce** [reply]
Hmmm, has to be more magical than 128..159

**Reply by Whammo** [reply]
possibly $AA and $55? (1010101 and 01010101)

**Reply by burnsauce** [reply]
No dice.  `(` and `stop` collide.

**Reply by Whammo** [reply]
Primes?

**Reply by burnsauce** [reply]
Using the 31 largest primes below 255 in sequence produces collisions just as quickly.

**Reply by Whammo** [reply]
EOR with the colliding string.

**Reply by Whammo** [reply]
But... you would have to keep track of that.... 1 128 ...

**Reply by Whammo** [reply]
Last char rshift lshift or 1 lshift lshift lshift...  Any operations that are reproduceable.

**Comment by Whammo** [comment]
Or go back to the collisions, pad them equally or inversely  'til they don't match anything anymore.

**Reply by Whammo** [reply]
Then there is not resetting the CRC for each word, or resetting every word, or every other word...
  Sorry, just brimming with bad ideas.

**Reply by burnsauce** [reply]
NO COLLUSION!

Sorry, couldn't help myself.

**Reply by Whammo** [reply]
What was it?

**Reply by burnsauce** [reply]
I haven't tried any of those yet.  I was making a joke about your typo.

**Comment by Whammo** [comment]
Typo? Heh!
